[Relay] Refactor Interpreter to treat lowering as IRModule->IRModule rewrite.

mbs-octoml opened the pull request:
This continues the work outlined in  
[TECompiler: Rewrite existing compile engine to match updated compiler flow](https://discuss.tvm.apache.org/t/rfc-relay-tecompiler-rewrite-existing-compile-engine-to-match-updated-compiler-flow/9233) to centralize the TE->TIR lowering logic into a single IRModule->IRModule transformation. This gets about halfway there for the Interpreter:
 - Remove direct access to TECompiler from interpreter, and instead call tec::LowerTEExpr when 'preparing' a module and expression for evaluation.
 - Make clear there's no phase distinction between create_interpreter and evaluate on the Python side -- both must be prepared together as a single IRModule.
 - But in return make sure the result of evaluate on the Python side is a packed func ready to directly apply 'simple' arguments to an already interpreted closure.
 - The interpreter builds and caches primitive TIR functions (and their corresponding dynamic shape functions) as packed funcs as they are encountered.
 - Clean uses of interpreter for constant folding on the C++ side.

Future work:
 - Fold LoweredModule into IRModule so tec::LowerTEExpr is just another pass.
 - Get rid of the implicit caching of lowered functions in TECompiler.
 - Make calling convention from Relay to TIR explicit, and remove all the function attribute hackery currently needed so the interpreter can correctly invoke lowered functions as it encounters them.
 - Make TECompiler private. Though could do this now it will make migrating the VM and AOT uses of CompilerEngine harder.

mbs-octoml commented:
@jroesch for feedback

jroesch commented on include/tvm/ir/module.h:
```diff
@@ -307,20 +307,31 @@ class IRModule : public ObjectRef {
-                                   const Map<GlobalTypeVar, TypeData>& type_definitions = {});
+                                   const Map<GlobalTypeVar, TypeData>& type_definitions = {},
+                                   std::unordered_set<String> import_set = {},
+                                   const std::string& name_hint = std::string());
```
Do we need to put this on the constructor? why not bind main here?

mbs-octoml replied:
Done. Kept FromExpr for the simple case (just an expr, always implicitly bound to 'main'). General case now FromExprInContext. Added yet another GetUniqueName into IRModule (though left private & not ffiable for now). We should bring the N copies of this under control.

jroesch commented:
cc @altanh

jroesch commented on python/tvm/relay/analysis/analysis.py:
```diff
@@ -433,8 +433,7 @@ def get_calibration_data(mod, data):
 
-    ref_ex = build_module.create_executor("graph", mod=mod, device=cpu(0))
-    ref_res = ref_ex.evaluate()(**data)
+    ref_res = build_module.create_executor("graph", mod=mod, device=cpu(0)).evaluate()(**data)
```
This is probably follow up, but this evaluate interface is sort of a hold over from the old days, we could probably just simplify the executor interface in another pass.

mbs-octoml replied:
Agree, and I'm very close to doing it now since I went and rejigged **all** the create_executor/evaluate calls anyway.

jroesch reviewed (approved):
I have a few nits, once those are fixed and CI are green you are good to go. Thanks so much for cleaning up this code.

mbs-octoml commented on src/ir/module.cc:
```diff
@@ -349,20 +349,23 @@ void IRModuleNode::Update(const IRModule& mod) {
   } else {
     func = relay::Function(relay::FreeVars(expr), expr, Type(), relay::FreeTypeVars(expr, mod), {});
   }
+  if (gv_name.empty()) {
```
Yeah, 'main' is endemic now. I got a bit of the way there addressing your FromExpr feedback above. Started to remove the dependence on 'main' but that got out of hand quickly so put it back in the bottle.

mbs-octoml commented on src/parser/parser.cc:
```diff
@@ -417,7 +417,7 @@ class Parser {
    */
   bool WhenMatch(const TokenType& token_type) {
-    DLOG(INFO) << "Parser::WhenMatch: Peek() == " << Peek();
+    // DLOG(INFO) << "Parser::WhenMatch: Peek() == " << Peek();
```
Ah, right! I started on per-file VLOGs, but hosted that out of here. Commented these out since I couldn't see the wood for the trees, will undo sorry.

mbs-octoml commented on src/relay/backend/interpreter.cc:
```diff
@@ -214,8 +217,12 @@ InterpreterState::InterpreterState(Expr current_expr, InterpreterState::Stack st
  public:
-  Interpreter(IRModule mod, Device device, Target target)
-      : mod_(mod), device_(device), target_(target), debug_op_(Op::Get("debug")) {}
+  Interpreter(IRModule mod, IRModule lowered_mod, Device device, Target target)
```
Done.

mbs-octoml commented on src/relay/backend/interpreter.cc:
```diff
@@ -283,73 +290,153 @@ class Interpreter : public ExprFunctor<ObjectRef(const Expr& n)>,
+   * \param all_prim_fn_vars Global vars for all lowered primitive functions needed by the above
+   * (including itself).
+   */
+  PackedFunc Build(const GlobalVar& prim_fn_var, const Array<GlobalVar>& all_prim_fn_vars) {
```
Done. PrimitiveToPackedFunc.

mbs-octoml commented on src/relay/backend/interpreter.cc:
```diff
@@ -283,73 +290,153 @@ class Interpreter : public ExprFunctor<ObjectRef(const Expr& n)>,
       auto arg = args[i];
-      auto param = func->params[i];
-      int state = cfunc->shape_func_param_states[i]->value;
+      int64_t state = prim_shape_fn_states[i]->value;
```
This kNeedInputData/kNeedInputShape state per arg is a bit gnarly since it's gathered as a side-effort of building the shape function. So I'm gonna leave a TODO to cleanup the way we convey this to call sites rather than kick that hornets nest right now.

Done for TupleType flattening.

**However** is ADT::Tuple coherent with FlattenTupleType? Ie is this legit?
```
std::vector<TensorType> result_tensor_types = FlattenTupleType(ftn->ret_type);
std::vector<ObjectRef> result_ndarrays;
for (int i = 0; i < result_tensor_types.size(); ++i) {
   ...
   result_ndarrays.emplace_back(...)
}
if (result_tensor_types.size() == 1) {
      return result_ndarrays[0];
} else {
  return ADT::Tuple(result_ndarrays);
}
```

mbs-octoml replied:
I ended up redoing the ADT vs NDArray stuff to mirror the FlattenTupleType and friends. Much cleaner. But for nested tuples there's going to be mismatches until we get this cleaned up throughout the codebase.

mbs-octoml commented on src/relay/backend/interpreter.cc:
```diff
@@ -359,32 +446,25 @@ class Interpreter : public ExprFunctor<ObjectRef(const Expr& n)>,
     size_t out_cnt = 0;
-    if (auto rtype = ret_type.as<TupleTypeNode>()) {
+    if (auto rtype = ftn->ret_type.as<TupleTypeNode>()) {
+      // TODO(mbs): Recursive flatten?
```
Done. Nice tetris effect of cascading simplifications, we should do this everywhere.

mbs-octoml commented on src/relay/backend/interpreter.cc:
```diff
@@ -476,17 +552,19 @@ class Interpreter : public ExprFunctor<ObjectRef(const Expr& n)>,
     TVMRetValue rv;
-    if (const TupleTypeNode* rtype = func->body->checked_type().as<TupleTypeNode>()) {
+    if (const TupleTypeNode* rtype = ftn->ret_type.as<TupleTypeNode>()) {
+      // TODO(mbs): Recursive flatten?
```
done.

altanh commented on tests/python/relay/test_backend_interpreter.py:
```diff
@@ -222,28 +218,72 @@ def test_tuple_passing():
+    y = relay.Var("y", relay.TensorType([n], "float32"))
+    mod = tvm.IRModule()
+    mod["add"] = relay.Function([x, y], relay.add(x, y))
+    x_np = np.random.uniform(size=(n)).astype("float32")
```
```suggestion
    x_np = np.random.uniform(size=(n,)).astype("float32")
```
fix single element tuple thing

altanh commented on tests/python/relay/test_backend_interpreter.py:
```diff
@@ -222,28 +218,72 @@ def test_tuple_passing():
+    mod = tvm.IRModule()
+    mod["add"] = relay.Function([x, y], relay.add(x, y))
+    x_np = np.random.uniform(size=(n)).astype("float32")
+    y_np = np.random.uniform(size=(n)).astype("float32")
```
```suggestion
    y_np = np.random.uniform(size=(n,)).astype("float32")
```

altanh commented on tests/python/relay/test_backend_interpreter.py:
```diff
@@ -222,28 +218,72 @@ def test_tuple_passing():
+    z = relay.add(x, y)
+    mod = tvm.IRModule()
+    mod["main"] = relay.Function([x, y], z)
+    x_np = np.random.uniform(size=(n)).astype("float32")
```
```suggestion
    x_np = np.random.uniform(size=(n,)).astype("float32")
```

altanh commented on tests/python/relay/test_backend_interpreter.py:
```diff
@@ -222,28 +218,72 @@ def test_tuple_passing():
+    mod = tvm.IRModule()
+    mod["main"] = relay.Function([x, y], z)
+    x_np = np.random.uniform(size=(n)).astype("float32")
+    y_np = np.random.uniform(size=(n)).astype("float32")
```
```suggestion
    y_np = np.random.uniform(size=(n,)).astype("float32")
```

altanh commented on include/tvm/ir/module.h:
```diff
@@ -307,20 +316,38 @@ class IRModule : public ObjectRef {
+  TVM_DLL static std::pair<IRModule, GlobalVar> FromExprInContext(
+      const RelayExpr& expr, const Map<GlobalVar, BaseFunc>& global_funcs = {},
+      const Map<GlobalTypeVar, TypeData>& type_definitions = {},
+      std::unordered_set<String> import_set = {});
```
just wondering about FFI interaction with `std::pair` and `std::unordered_set`, could there be problems?

mbs-octoml replied:
Good point -- the ffi binding hides both of them but let me find a Better Way.

mbs-octoml replied:
Since it's unclear whether the python side needs the global var or import set I removed the TVM_DLL and put FromExpr back how it was as the ffi-clean way.

mbs-octoml commented:
@jroesch PTAL:
 - BUG: I was compiling dynamic shape functions for the target even though they must always stay on the cpu. Fixed, but it means a bit more bookkeeping in Interpreter.
 - Cleaned up flattening of tuples, both for TupleType/TensorType and runtime::ADT/runtime::NDArray.
 Thanks!

jroesch commented on src/relay/backend/interpreter.cc:
```diff
@@ -39,9 +42,81 @@
+
+// Analogue of FlattenTupleType for runtime ADT vs NDArray values.
+// TODO(mbs): Hoist somewhere sensible, maybe op/memory.h?
+void FlattenADTAux(const ObjectRef& object_ref, std::vector<NDArray>* out) {
```
Yeah can we maybe just add an issue to unify these.

mbs-octoml replied:
done

jroesch commented on src/relay/backend/interpreter.cc:
```diff
@@ -283,251 +361,301 @@ class Interpreter : public ExprFunctor<ObjectRef(const Expr& n)>,
+   * will be the interpreter's target_. However for shape functions this will be the generic
+   * 'cpu' target, since shape functions are always executed on the host cpu.
+   */
+  PackedFunc TIRToPackedFunc(const GlobalVar& tir_fn_var, const Array<GlobalVar>& all_tir_fn_vars,
```
Is this useful outside for the interpreter?

mbs-octoml replied:
Almost certainly, figured as we pull on the build loose ends this will sort itself out.

jroesch commented on src/relay/backend/te_compiler.cc:
```diff
@@ -408,27 +462,135 @@ class LowerTensorExpr : public ExprMutator {
+      // TODO(mbs): Capture all this as part of a 'call into TIR' construct once available.
+      // The way the shape function calling convention is derived and passed to call sites
+      // via the 'parameter states' could be improved.
+      tir_call_attrs->metadata.Set("prim_shape_fn_var", lowered_shape_func->prim_fn_var);
```
Possible follow up, maybe we can refactor into composite structure in the Attrs?

mbs-octoml replied:
Yeah, I think this should be a struct inside the 'call-tir' op.

jroesch commented on src/relay/backend/te_compiler_cache.h:
```diff
@@ -213,6 +213,7 @@ CachedFunc PrimFuncFor(const Function& source_func, const Target& target,
 CachedFunc ShapeFuncFor(const Function& prim_func, const Target& target,
                         std::function<std::string(std::string)> renamer);
 
+// TODO(mbs): Bring name uniqification under control -- this is replicated in quite a few places.
```
Great, probably worth adding ticket for all the follow up.

mbs-octoml replied:
done.

jroesch reviewed (approved):
LGTM, just some stuff we should do as follow up.

electriclilies reviewed (commented):
Thanks for all your hard work Mark! The PR looks good to me so far, I really appreciate the high-quality comments and documentation. I'm still working my way through because it is quite long :)

electriclilies reviewed (approved):
LGTM! Thanks again :)
